### ISSUE TITLE
409 conflict response from AWS SCIM API when syncing large number of users

### ISSUE THREAD

**Issue by hadrianbs** [awslabs/ssosync#65]
**Describe the bug**
I've got this issue where I keep getting `409 conflict` response from AWS SCIM API when syncing ~100 users

I did some reading to the code and AWS SCIM API reference, and found this on the ListUsers page:

https://docs.aws.amazon.com/singlesignon/latest/developerguide/listusers.html#constraints-listusers

> At this time, the ListUsers API is only capable of returning up to 50 results

I think this is causing the issue. The current implementation uses ListUsers to fetch AWS users, but since its only capable of returning only 50 users, it will only compare the 50 AWS users against the ~100 from Google Workspace API. This is causing 409 conflicts to AWS SCIM when ssosync attempts to create an already existing AWS user.

Is there anything I miss when configuring ssosync?

**Comment by hazelguo** [comment]
Same problem! any updates or any workaround?

**Comment by hadrianbs** [comment]
@hazelguo ended forking and modified the AWS SSO user creation part. I added a check before creating the user on AWS SSO side, if the user already exist, then it will skip the user creation.
This causes a lot more API calls to AWS SCIM API, but solves the 409 conflict problem.
I'm planning to open a PR to address this issue when I got the time,  hopefully soon.

**Comment by christiangda** [comment]
@hadrianbs @hazelguo there is no way to mitigate this using this program (at least without a big refactoring) and this is one of the reasons that motivate me to find a different alternative.

Since some time ago I have been working on a new approach of the sync, today I have been tested my last code in the `devel branch` here [https://github.com/slashdevops/idp-scim-sync/tree/develop](https://github.com/slashdevops/idp-scim-sync/tree/develop).  I will deliver the version v0.0.1 soon, as soon I write some documentation.

for the new approach, I'm using the same method that `terraform` use, an `state file` to avoid the use of the AWS SSO SCIM API calls after the first sync.  What I mean is, once the first sync occurs, the following sync only compare changes with this state file stored in `AWS S3` with mean only one `AWS S3 API call` to get all the data, and once the comparison exists with the fresh data coming from `GWS` the program only perform a specific call to `AWS SSO SCIM API` with the differences.

If you want to see what this is about, check the images files here [https://github.com/slashdevops/idp-scim-sync/tree/develop/docs/images/demo](https://github.com/slashdevops/idp-scim-sync/tree/develop/docs/images/demo) I created these today.

WARNING: if do you want to test it, just `clone the project` and use the `develop branch` execute `make` command.

then `./build/idpscim --help` 

```bash
./build/idpscim --help

Sync your Google Workspace Groups and Users to AWS Single Sing-On using
AWS SSO SCIM API (https://docs.aws.amazon.com/singlesignon/latest/developerguide/what-is-scim.html).

Usage:
  idpscim [flags]

Flags:
  -k, --aws-s3-bucket-key string          AWS S3 Bucket key to store the state
  -b, --aws-s3-bucket-name string         AWS S3 Bucket name to store the state
  -t, --aws-scim-access-token string      AWS SSO SCIM API Access Token
  -e, --aws-scim-endpoint string          AWS SSO SCIM API Endpoint
  -d, --debug                             fast way to set the log-level to debug
  -s, --gws-service-account-file string   path to Google Workspace service account file (default "credentials.json")
  -u, --gws-user-email string             Google Workspace user email with allowed access to the Google Workspace Service Account
  -h, --help                              help for idpscim
  -f, --log-format string                 set the log format (default "text")
  -l, --log-level string                  set the log level [panic|fatal|error|warn|info|debug|trace] (default "info")
  -q, --query-groups strings              Google Workspace Groups query parameter, example: --query-groups 'name:Admin* email:admin*' --query-groups 'name:Power* email:power*', see: https://developers.google.com/admin-sdk/directory/v1/guides/search-groups
  -r, --query-users strings               Google Workspace Users query parameter, example: --query-users 'name:Admin* email:admin*' --query-users 'name:Power* email:power*', see: https://developers.google.com/admin-sdk/directory/v1/guides/search-users
  -n, --state-enabled                     enable state
  -m, --sync-method string                Sync method to use [groups] (default "groups")
  -v, --version                           version for idpscim
```

**Comment by ChrisPates** [comment]
This issue was addressed in v2.0 of SSOSync.
